include/envoy/grpc/async_client.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <chrono>
+
+#include "envoy/common/optional.h"
+#include "envoy/common/pure.h"
+#include "envoy/http/header_map.h"
+
+namespace Envoy {
+namespace Grpc {
+
+/**
+ * Async client failure reasons.
+ */
+enum class AsyncClientFailureReason {
+  // The stream has been reset.
+  Reset,
+  // The stream has gRPC errors.
+  // TODO(htuch): Consider providing more granular gRPC failure information (e.g. status code,
+  // framing error, resource exhaustion, authentication failure).
+  GrpcError,
+};
+
+/**
+ * An in-flight gRPC stream.
+ */
+template <class RequestType> class AsyncClientStream {
+public:
+  virtual ~AsyncClientStream() {}
+
+  /**
+   * Send request message to the stream.
+   * @param request protobuf serializable message.
+   * @param end_stream supplies whether this is the last message.
+   */
+  virtual void sendMessage(const RequestType& request, bool end_stream) PURE;
```

Comment:
You probably need a method to just send end_stream (empty DATA frame with end_stream)